S3 설정 및 브랜드 이미지 구현

Hosick opened the pull request:
- S3 설정 및 Service 구현
- 업로드 할 파일명 난수로 변경 구현
- 업로드 할 파일 mimeType 검사 구현
- 브랜드 이미지 업로드 구현
- 브랜드 이미지 없이 저장 분리

Hosick committed:
feat: S3 설정 및 브랜드 이미지 저장

-S3 설정 및 Service 구현
-업르드 할 파일명 난수로 변경 구현
-업로드 할 파일 mimeType 검사 구현
-브랜드 이미지 업로드 구현
-브랜드 이미지 없이 저장 분리

issue number: #35

Hosick committed:
refactor: contentType 리팩토링

- 이미지 파일 타입 체크 리팩토링
- 업로드를 위한 metadata에 contentType 추가

issue number: #35

Hosick committed:
refactor: 브랜드 이미지 업로드 리팩토링

- image 포함 검사 책임을 controller에서 service로 변경

issue number: #35

ROMANIAPEOPLE commented on src/main/java/com/flab/shoeauction/service/BrandService.java:
```diff
@@ -14,18 +16,26 @@
         if (checkDuplicateName(requestDto)) {
             throw new DuplicateBrandNameException();
         }
+        if (brandImage != null) {
+            String imagePath = awsS3Service.upload(brandImage, BRAND_IMAGES_DIR);
+            requestDto.setImagePath(imagePath);
+        }
```
이부분 궁금한게 있는데 브랜드나 상품을 등록할때 이미지 없이 등록해야 할 일이 있을까요???

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/BrandApiController.java:
```diff
@@ -30,8 +32,9 @@
+    public ResponseEntity<Void> createBrand(@Valid @RequestPart SaveRequest requestDto,
+        @RequestPart(required = false) MultipartFile brandImage) {
+        brandService.saveBrand(requestDto, brandImage);
         return CREATED;
```
이거는 `@ResponseStatus`로 생략이 가능합니다

Hosick replied:
몰랐던 방법 알려주셔서 감사합니다! 적용하도록 하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/exception/file/IllegalMimeTypeException.java:
```diff
@@ -0,0 +1,5 @@
+package com.flab.shoeauction.exception.file;
+
+public class IllegalMimeTypeException extends IllegalArgumentException{
```
커밋전에는 컨벤션 체크를 해주세요. pre commit hook을 쓰셔도 좋습니다~

```suggestion
public class IllegalMimeTypeException extends IllegalArgumentException{
```

Hosick replied:
checkstyle pre commit hook을 적용하려고 하는 중인데 쉘 스크립트 짜는 것이 익지 않아서 시간이 조금 걸리는 것 같습니다..
적용 전까지는 커밋할 때 스타일 체크하도록 노력하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/AwsS3Service.java:
```diff
@@ -0,0 +1,73 @@
+
+    private AmazonS3 s3Client;
+
+    @Value("${cloud.aws.credentials.accessKey}")
+    private String accessKey;
+
+    @Value("${cloud.aws.credentials.secretKey}")
+    private String secretKey;
+
+    @Value("${cloud.aws.s3.bucket}")
+    private String bucket;
+
+    @Value("${cloud.aws.region.static}")
+    private String region;
```
이런것도 생성자 주입이 가능합니다

Hosick replied:
Value로 주입받는 것보다 생성자로 한 번에 주입받는게 깔끔할 것 같습니다. 반영하겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/AwsS3Service.java:
```diff
@@ -0,0 +1,73 @@
+                new PutObjectRequest(bucket, fileNameWithPath, file.getInputStream(), metadata)
+                    .withCannedAcl(CannedAccessControlList.PublicRead));
+        } catch (IOException exception) {
+            exception.printStackTrace();
```
이거는 절대 쓰면 안돼는 메소드입니다~
심지어 업로드가 실패했음에도 로그를 찍고나서는 정상적인 코드 흐름으로 처리하네요

Hosick replied:
넵 토비의 스프링에서도 stackTrace만 찍고 넘어가는 방법은 매우 안좋다고 보았습니다!
다만 예외 처리가 고민되는 부분이라 여쭤보려고 이렇게 두었는데, 
ImageRoadFailedException 정도의 커스텀 예외로 전환하여 처리하는 것이 적절하다고 판단되어 수정하겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/file/FileService.java:
```diff
@@ -0,0 +1,42 @@
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
```
이 클래스는 의존관계도 갖지 않고 확장의 여지도 없어보입니다. 유틸클래스로 만들어줘도 괜찮을 것 같네요

Hosick replied:
말씀하신 이유대로 유틸클래스로 변경하는 것이 맞는 것 같습니다! 변경하도록 하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/file/FileService.java:
```diff
@@ -0,0 +1,42 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
```
이거는 습관성 복붙으로 보이는데 이 어노테이션은 꼭 필요한가요?

Hosick replied:
이 부분 직접 타이핑했습니다....😢

원래 구현 시 파일명을 변경할 때 `encrytionService`를 주입받아서 암호화로 하려고 하다가,
굳이 파일명을 복호화할 필요도 없는데 랜덤 문자열로 사용하는게 더 나을 것 같아서 변경했는데
생성자 제거를 깜빡한 것 같습니다. 바로 수정하겠습니다~

Hosick committed:
refactor: 리뷰 반영

- ResponseStatus 적용
- Aws Properties 생성자 주입으로 변경
- FileService를유틸 클래스로 변경
- 이미지 IOExcpetion 예외 전환 구현

issue number: #35

f-lab-dev commented on src/main/java/com/flab/shoeauction/exception/GlobalExceptionHandler.java:
```diff
@@ -151,4 +153,11 @@ public final ResponseEntity handleNotAuthorizedException(NotAuthorizedException
+    @ExceptionHandler(ImageRoadFailedException.class)
+    public final ResponseEntity<String> imageRoadFailedException(
+        ImageRoadFailedException ex) {
+        log.debug("이미지를 로드에 실패하였습니다.", ex);
```
`이미지를 로드하는데에 실패하였습니다` 가 맞는 표현일 것 같아요~
혹은 `이미지 로드에 실패하였습니다`

Hosick replied:
문구 사용할 때도 조금 더 생각해보면서 작성하겠습니다!

f-lab-dev reviewed (approved):
코멘트드린 하나만 수정하고 머지하셔도 될 것 같네요~

Hosick committed:
chore: 리뷰 반영

- 로그 수정

issue number: #35
